marshal/marshal.go

```diff
@@ -0,0 +1,212 @@
+		tag, err := parseJSONTag(jsonTag)
+		if err != nil {
+			return nil, err
+		}
+
+		fieldValue := value.Field(i)
+		fieldType := objectType.Field(i)
+
+		// Anonymous fields should be marshalled using the same JSON, and then merged into the same map
+		if fieldType.Anonymous && fieldValue.IsValid() {
+			anonymousFieldResult, err := structAsMap(fieldValue.Interface())
+			if err != nil {
+				return nil, err
+			}
+			result = mergeMaps(anonymousFieldResult, result)
+			continue
+		}
+
+		// Skip fields which should not be included
+		if !includeField(tag, fieldValue, includeFields[fieldType.Name]) {
+			continue
+		}
+
+		if tag.asString {
+			result[tag.name] = formatAsString(fieldValue, fieldType.Type.Kind())
+		} else {
+			result[tag.name] = fieldValue.Interface()
+		}
+	}
+	return result, nil
+}
+
+// returns the element as string
+func formatAsString(v reflect.Value, kind reflect.Kind) string {
+	if kind == reflect.Ptr && !v.IsNil() {
+		v = v.Elem()
+	}
+
+	return fmt.Sprintf("%v", v.Interface())
+}
+
+type jsonTag struct {
+	name      string
+	asString  bool
+	ignore    bool
+	omitempty bool
+}
+
+func parseJSONTag(raw string) (jsonTag, error) {
+	if raw == "-" || raw == "" {
+		return jsonTag{ignore: true}, nil
+	}
+
+	parts := strings.Split(raw, ",")
+
+	jsonTag := jsonTag{
+		name: parts[0],
+	}
+
+	for _, v := range parts {
+		switch v {
+		case "omitempty":
+			jsonTag.omitempty = true
+		case "string":
+			jsonTag.asString = true
+		}
+	}
+	return jsonTag, nil
+}
+
+// Determines whether a field should be included or not
+func includeField(tag jsonTag, value reflect.Value, forceSend bool) bool {
+	if tag.ignore {
+		return false
+	}
+	if !tag.omitempty {
+		return true
+	}
+	return (isBasicType(value.Type()) && forceSend) || !isEmptyValue(value)
+}
+
+// isEmptyValue returns whether v is the empty value for its type.
+// This implementation is based on on the encoding/json package for consistency on the results
+// https://github.com/golang/go/blob/a278550c40ef3f01a5fcbef43414dc49009201f8/src/encoding/json/encode.go#L306
+func isEmptyValue(v reflect.Value) bool {
+	switch v.Kind() {
+	case reflect.Array, reflect.Map, reflect.Slice, reflect.String:
+		return v.Len() == 0
+	case reflect.Bool:
+		return !v.Bool()
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return v.Int() == 0
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		return v.Uint() == 0
+	case reflect.Float32, reflect.Float64:
+		return v.Float() == 0
+	case reflect.Interface, reflect.Pointer:
+		return v.IsNil()
+	}
+	return false
+}
+
+// Merges two maps. Overrides duplicated elements, which is fine because json.Unmarshal also
+// does it, and a JSON should not have duplicated entries.
+func mergeMaps(m1 map[string]any, m2 map[string]any) map[string]any {
+	merged := make(map[string]any)
+	maps.Copy(merged, m2)
+	maps.Copy(merged, m1)
```

Comment:
This order means m1 has precedence over m2. From the argument order of the function I would have expected the reverse, i.e. for m2 to have precedence, because it is the second argument.